File: exercises/exercises-mpi-mpl/stridesend.cxx
/* %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%
   %%%% This program file is part of the book and course
   %%%%   "Parallel Computing for Science and Engineering"
   %%%% by Victor Eijkhout, copyright 2020
   %%%%
   %%%% MPI example for vector type using MPL
   %%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
*/

#include <iostream>
#include <sstream>
using namespace std;
#include <mpl/mpl.hpp>

int main(int argc,char **argv) {

  const mpl::communicator &comm_world = mpl::environment::comm_world();
  int nprocs,procno;
  // compute communicator rank and size
  nprocs = comm_world.size();
  procno = comm_world.rank();

  stringstream proctext;

  if (nprocs<2) {
    proctext << "This program needs at least two processes" << endl;
    cerr << proctext.str(); proctext.clear();
    return -1;
  }
  int sender = 0, localsize = 10;

  if (procno==sender) {

    /*
     * Create big data array to be sent
     */
    int ndata = localsize*nprocs;
    int *data = (int*) malloc(ndata*sizeof(int));
    if (!data) {
      proctext << "Out of memory" << endl; 
      cerr << proctext.str(); proctext.clear();
      comm_world.abort(0); }
    for (int i=0; i<ndata; i++)
      data[i] = i;

    /*
     * Exercise
     * - you need a datatype for sending data. Can you define it here....
     */
    int count,stride,blocklength;
/**** your code here ****/

    for (int sendto=0; sendto<nprocs; sendto++) {
      if (sendto==procno)
	continue;
      /*
       * - or do you define the datatype here?
       * - then do a send to the other processor.
       */
/**** your code here ****/
    }
  } else {
    int *mydata = (int*) malloc(localsize*sizeof(int));
    comm_world.recv(mydata,mpl::contiguous_layout<int>(localsize),sender);
    for (int i=0; i<localsize; i++)
      if (mydata[i]%nprocs!=procno) {
	proctext << "[" << procno << "] received element=" << mydata[i]
		 << ", should be " << i*nprocs+procno << endl;
	cerr << proctext.str(); proctext.clear();
      }
  }

  if (procno==0) {
    proctext << "Finished" << endl;
    cerr << proctext.str(); proctext.clear();
  }

  return 0;
}

File: exercises/exercises-mpi-p/stridesend.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
####   Parallel Computing for Science and Engineering
#### by Victor Eijkhout, copyright 2013-8
####
#### stridesend.py : MPI python exercise for vector type
####
################################################################
################################################################

from mpi4py import MPI
import numpy as np
import sys

comm = MPI.COMM_WORLD

nprocs = comm.Get_size()
procno = comm.Get_rank()

if nprocs<2:
    print("C'mon, get real....")
    sys.exit(1)

## process zero will do the sending
sender = 0
## how many elements to each process?
count = 6

## allocate source data only on the sender
if procno==sender:
    source = np.empty(nprocs*count,dtype=np.intc)
    for i in range(nprocs*count):
        source[i] = i
else: source = None

## everyone gets a target array
target = np.zeros(count,dtype=np.intc)

##
## Now set up the vector type and send to everyone but yourself
##
if procno==sender:
    ##
    ## Make a vector type for sending `count' strided elements
    ## Important hint: the same type can be used for all sends
    ##
    vectortype = MPI.INT.Create_vector(
#### your code here ####
    )
    vectortype.Commit()
    for p in range(nprocs):
        ##
        ## Do not send to yourself
        ## Exercise: how would you send to yourself
        ##     without risk of deadlock?
        ##
        if p==sender:
            continue
        ##
        ## Send to process p
        ## Exercise: how do you determine what data to send?
        ##
        comm.Send(
            [
                ##
                ## Fill in the buffer/count/datatype triple
                ## Note: buffer needs to be an array !!!
                ##
#### your code here ####
                ],
            p
            )
    vectortype.Free()
else:
    ##
    ## If you are not the sender, receive data and check correctness
    ##
    comm.Recv([target,count,MPI.INT],0)
    ## disabled test
    for i in range(count):
        if target[i]!=nprocs*i+procno:
            print("Mismatch: proc %d, elt %d is %d s/b %d" % \
                  (procno,i,target[i],nprocs*i+procno))

if procno==0:
    print("Finished")
